src/trace/implementations/merge_batcher_col.rs

```diff
@@ -214,116 +35,89 @@ impl<D: Ord+Columnation+'static, T: Ord+Columnation+'static, R: Semigroup+Column
+    /// Helper to get pre-sized vector from the stash.
+    #[inline]
+    fn empty(&self, stash: &mut Vec<TimelyStack<T>>) -> TimelyStack<T> {
+        stash.pop().unwrap_or_else(|| TimelyStack::with_capacity(self.preferred_buffer_size()))
     }
 
-    /// Insert an empty buffer into the stash. Panics if the buffer is not empty.
-    fn recycle(&mut self, mut buffer: TimelyStack<(D, T, R)>) {
-        if buffer.capacity() == Self::buffer_size() && self.stash.len() < 2 {
-            buffer.clear();
-            self.stash.push(buffer);
+    /// Helper to return a batch to the stash.
+    #[inline]
+    fn recycle(&self, mut batch: TimelyStack<T>, stash: &mut Vec<TimelyStack<T>>) {
+        // TODO: Should we limit the size of `stash`?
+        if batch.capacity() == self.preferred_buffer_size() /*&& stash.len() < 2*/ {
+            batch.clear();
+            stash.push(batch);
         }
     }
+}
+
+impl<K, V, T,R> Merger for ColumnationMerger<((K,V), T, R)>
+where
+    K: Columnation + Ord + Data,
+    V: Columnation + Ord + Data,
+    T: Columnation + Ord + PartialOrder + Data,
+    R: Columnation + Semigroup + 'static,
+{
+    type Time = T;
+    type Input = Vec<((K,V),T,R)>;
+    type Batch = TimelyStack<((K,V),T,R)>;
+    type Output = ((K,V),T,R);
 
-    fn push(&mut self, batch: &mut Vec<(D, T, R)>) {
+    fn accept(&mut self, batch: RefOrMut<Self::Input>, stash: &mut Vec<Self::Batch>) -> Vec<Self::Batch> {
+        // `batch` is either a shared reference or an owned allocations.
+        let mut batch: Vec<_> = match batch {
+            RefOrMut::Ref(vec) => {
+                let mut owned = Vec::default();
+                owned.clone_from(vec);
+                owned
+            }
+            RefOrMut::Mut(vec) => std::mem::take(vec)
+        };
         // Ensure `self.pending` has a capacity of `Self::pending_buffer_size`.
-        if self.pending.capacity() < Self::pending_buffer_size() {
-            self.pending.reserve(Self::pending_buffer_size() - self.pending.capacity());
+        if self.pending.capacity() < self.pending_buffer_size() {
+            self.pending.reserve(self.pending_buffer_size() - self.pending.capacity());
         }
 
+        let mut output = Vec::default();
+
         while !batch.is_empty() {
             self.pending.extend(batch.drain(..std::cmp::min(batch.len(), self.pending.capacity() - self.pending.len())));
             if self.pending.len() == self.pending.capacity() {
-                crate::consolidation::consolidate_updates(&mut self.pending);
+                consolidate_updates(&mut self.pending);
                 if self.pending.len() > self.pending.capacity() / 2 {
                     // Flush if `self.pending` is more than half full after consolidation.
-                    self.flush_pending();
+                    let mut stack = self.empty(stash);
+                    stack.reserve_items(self.pending.iter());
+                    for tuple in self.pending.drain(..) {
```

Comment:
Perhaps we want to copy less than everything, as this may produce irregularly shaped batches?